### DISCUSSION TITLE
제네릭(generic)

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#36] [프론트 기술 질문 공간] [answered]
기억에 남는 제네릭 사용경험이 있다면 설명해주세요. 없다면 제네릭에 대해 알고 있는대로 설명해주세요.

**Comment by javascriptbiggosoo** [comment]
타입스크립트에서 제네릭을 이용하면 함수를 정의할 때에 argument의 타입을 이어받아 함수 내에서 사용하도록 해줍니다. 사용범위가 넓은 함수에서 주로 쓰이는데 본인이 별로 제네릭을 사용하지 않는 개발자더라도 사용하는 오픈소스 라이브러리에서 흔히 볼 수 있습니다.

**Comment by michoball** [comment]
함수를 정의할 때 함수에 사용되는 타입을 특정하지 않고 사용하고 싶을 때 사용합니다. any와는 다르게 어떤 함수를 사용하였는지 명확히 할 수 있습니다. 
범용적으로 사용되는 함수를 정의할 때 주로 사용합니다. 
제네릭을 이용한 프로젝트를 구현한 적은 없는 것 같습니다. 하지만 typescript 관련 문제 풀면서 사용해보며 학습하고 있습니다. 

ex ) https://typescript-exercises.github.io/#exercise=7&file=%2Findex.ts

**Comment by bona373737** [comment]
정적 타입 언어에서는 함수 및 클래스를 선언하는 시점에서 매개변수 혹은 리턴 타입을 정의해야하기 때문에 기본적으로는 특정 타입을 위해 만들어진 클래스나 함수를 다른 타입을 위해 재사용할 수가 없다. 

제네릭을 사용하면 클래스 혹은 함수에서 사용할 타입을 그 함수나 클래스를 사용할 때 지정할 수 있다.

**Comment by WooGie911** [comment]
딱 기억에 남는 사용 경험은 없고 제네릭에 대해 알고 있는대로 설명해보겠습니다.

제네릭(generic)이란 데이터의 타입(data type)을 일반화한다(generalize)는 것을 의미합니다.

타입스크립트에서 제네릭을 이용하면 함수의 타입을 정의 할 때 arg의 타입을 이어받아 사용 할 수 있습니다.

즉, 어떤 정보도 잃지 않고 받아서 사용 할 수 있습니다.

**Comment by BaekYuri** [comment]
자바스크립트에서 사용한 적은 없지만, 자바에서 사용해보았던 경험이 있습니다.
List, Set, Map을 사용할 때, Comparator를 이용해 sort 기능을 구현할 때 사용해보았습니다.
예를들어 List<String> strList 라는 객체를 정렬시키기 위해 new Comparator<String>을 하여 Comparator 객체를 생성하고, implements 되어있는 함수인 compare 함수를 구현하였던 경험이 있습니다.

**Comment by skyhanull** [comment]
제네릭이란 타입을 마치 함수의 파라미터처럼 사용하는 것을 의미합니다 즉 Generic은 코드를 작성할 때가 아니라 코드를 수행될 때(런타임) 타입을 명시하기에  여러 가지 타입에서 동작하는 컴포넌트를 생성하는 데 사용됩니다.
실제로 사용해본 경험은 없습니다.

**Comment by WooGie911** [comment, marked as answer]
제네릭(generic)이란 데이터의 타입(data type)을 일반화한다(generalize)는 것을 의미합니다.

정적 타입 언어에서는 함수 및 클래스를 선언하는 시점에서 매개변수 혹은 리턴 타입을 정의해야하기 때문에 
기본적으로는 특정 타입을 위해 만들어진 클래스나 함수를 다른 타입을 위해 재사용할 수가 없습니다.

그래서 제네릭을 사용하여 클래스 혹은 함수에서 사용할 타입을 그 함수나 클래스를 사용할 때 지정할 수 있습니다.

또한 타입스크립트에서 제네릭을 이용하면 함수의 타입을 정의 할 때 arg의 타입을 이어받아 사용 할 수 있습니다.

즉, 어떤 정보도 잃지 않고 받아서 사용 할 수 있으며 any와는 다르게 어떤 함수를 사용하였는지 명확히 할 수 있습니다.
